mlflow/pyfunc/__init__.py

```diff
@@ -760,6 +902,40 @@ def _warn_potentially_incompatible_py_version_if_necessary(model_py_version=None
             model_py_version,
             PYTHON_VERSION,
         )
+        return True
+    return False
+
+
+def _get_or_create_model_cache_dir():
+    nfs_root_dir = get_nfs_cache_root_dir()
+    if nfs_root_dir is not None:
+        # In databricks, the '/local_disk0/.ephemeral_nfs' is mounted as NFS disk
+        # the data stored in the disk is shared with all remote nodes.
+        root_dir = os.path.join(nfs_root_dir, "models")
+        os.makedirs(root_dir, exist_ok=True)
+        tmp_model_dir = tempfile.mkdtemp(dir=root_dir)
+        # TODO: register deleting tmp_model_dir handler when exit
+    else:
+        import atexit
+        import shutil
+        tmp_model_dir = tempfile.mkdtemp()
+        atexit.register(shutil.rmtree, tmp_model_dir, ignore_errors=True)
+
+    return tmp_model_dir
+
+
+def _create_symlink(src, dst):
+    # The logic here is for addressing some race conditions.
+    try:
+        if os.path.exists(dst) and os.readlink(dst) != src:
+            os.remove(dst)
+    except:
+        pass
+
+    try:
+        os.symlink(src, dst)
+    except:
+        pass
 
 
 def spark_udf(spark, model_uri, result_type="double"):
```

Comment:
We should introduce an argument (e.g. `env=...`) for controlling the environment that is used for UDF inference

1. `local` (default)
2. `conda`
3. `virtualenv`

For MLflow 1.x, `local` should remain the default. For MLflow 2.0, we can make `virtualenv` the default.